qiskit/transpiler/passes/utils/gates_basis.py

```diff
@@ -18,22 +18,37 @@
 class GatesInBasis(AnalysisPass):
     """Check if all gates in a DAG are in a given set of gates"""
 
-    def __init__(self, basis_gates):
+    def __init__(self, basis_gates, target=None):
         """Initialize the GatesInBasis pass.
 
         Args:
             basis_gates (list): The list of strings representing the set of basis gates.
+            target (Target): If specified the target represnting the backend. If specified
+                this will be used instead of the ``basis_gates`` parameter
         """
         super().__init__()
         self._basis_gates = set(basis_gates)
+        self._target = target
 
     def run(self, dag):
         """Run the GatesInBasis pass on `dag`."""
         gates_out_of_basis = False
-        basic_instrs = {"measure", "reset", "barrier", "snapshot", "delay"}
-        for gate in dag._op_names:
-            if gate not in self._basis_gates and gate not in basic_instrs:
-                gates_out_of_basis = True
-                break
-
+        if self._target is not None:
+            qubit_map = {qubit: index for index, qubit in enumerate(dag.qubits)}
+            for gate in dag.topological_op_nodes():
```

Comment:
I need to benchmark whether this or `dag.op_nodes()` is faster and potentially adjust this based on that. I did it this way because I thought this would be faster (but that was because I misremembered the return type from retworkx)

Reply:
I did this earlier and `dag.op_nodes()` is noticeably faster, ~2x so. So I changed to that in https://github.com/Qiskit/qiskit-terra/pull/7548/commits/9f9c9351ea6dd26346a89d1d3c0f08de28f9b54e